oak/server/rust/oak_runtime/src/node/roughtime.rs

```diff
@@ -16,32 +16,170 @@
 
-use crate::runtime::RuntimeProxy;
-use log::{info, warn};
+use crate::{
+    io::Receiver,
+    node::grpc::invocation::Invocation,
+    runtime::RuntimeProxy,
+    time::{
+        get_default_servers, RoughtimeClient, RoughtimeServer, DEFAULT_MAX_RADIUS_MICROSECONDS,
+        DEFAULT_MIN_OVERLAPPING_INTERVALS, DEFAULT_SERVER_RETRIES, DEFAULT_TIMEOUT_SECONDS,
+    },
+};
+use log::{debug, error, info};
+use oak_abi::{
+    proto::{
+        google::rpc::Code,
+        oak::{
+            application::RoughtimeClientConfiguration,
+            encap::GrpcResponse,
+            roughtime::{RoughtimeRequest, RoughtimeResponse},
+        },
+    },
+    OakStatus,
+};
+use prost::Message;
 use tokio::sync::oneshot;
 
 /// Roughtime client pseudo-Node.
 pub struct RoughtimeClientNode {
     node_name: String,
+    client: RoughtimeClient,
 }
 
 impl RoughtimeClientNode {
     /// Creates a new [`RoughtimeClientNode`] instance, but does not start it.
-    pub fn new(node_name: &str) -> Self {
+    pub fn new(node_name: &str, config: &RoughtimeClientConfiguration) -> Self {
+        let timeout_seconds = if config.timeout_seconds == 0 {
+            DEFAULT_TIMEOUT_SECONDS
+        } else {
+            config.timeout_seconds as u64
+        };
+        let server_retries = if config.server_retries == 0 {
+            DEFAULT_SERVER_RETRIES
+        } else {
+            config.server_retries as usize
+        };
+        let min_overlapping_intervals = if config.min_overlapping_intervals == 0 {
+            DEFAULT_MIN_OVERLAPPING_INTERVALS
+        } else {
+            config.min_overlapping_intervals as usize
+        };
+        let max_radius_microseconds = if config.max_radius_microseconds == 0 {
+            DEFAULT_MAX_RADIUS_MICROSECONDS
+        } else {
+            config.max_radius_microseconds
+        };
+        let servers = if config.servers.is_empty() {
+            get_default_servers()
+        } else {
+            config
+                .servers
+                .iter()
+                .map(|server| {
+                    RoughtimeServer::new(
+                        &server.name,
+                        &server.host,
+                        server.port as u16,
+                        &server.public_key_base64,
+                    )
+                })
+                .collect()
+        };
+
         Self {
             node_name: node_name.to_string(),
+            client: RoughtimeClient::new(
+                servers,
+                min_overlapping_intervals,
+                max_radius_microseconds,
+                timeout_seconds,
+                server_retries,
+            ),
         }
     }
+
+    /// Processes a gRPC invocation.
+    fn process_invocation(&self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        match invocation.receive_request(runtime) {
+            Ok(request) => {
+                if request.method_name != "/oak.roughtime.RoughtimeService/GetRoughtime" {
```

Comment:
Should this be auto generated at some point? Do we have a GitHub issue to track that already?

Reply:
Created #1113